initial work on 0.9 blog post (#101)
AvaloniaUI/avaloniaui.net: wwwroot/blog/2019-12-13-avalonia-0.9.0-release.md

Title: Avalonia 0.9.0 Release
Published: 2019-12-13
Category: Release
Author: Dan Walmsley
---

We are pleased to announce that [Avalonia](https://github.com/AvaloniaUI/Avalonia) 0.9.0 has been
released.

0.9 is a huge update, it has been extensively tested and brings some great new features and improvements.

# Compiled XAML

> In previous releases XAML was parsed at runtime. This meant that some syntax errors could cause runtime crashes, and may be hard to identify.

Reviewer: I think perf is probably the most important part of the XAML compiler. In addition, it wasn't just syntax changes that caused crashes, also misspelled properties, missing classes etc

```suggestion
In previous releases XAML was parsed at runtime. This had a non-significant runtime overhead and meant that some errors could cause runtime crashes.
```

> Now we have a scratch built XAML compiler that parses all the XAML at build time, and converts it directly to MSIL.

Reviewer: ```suggestion
For 0.9 we have built a XAML compiler that parses XAML at build time, and compiles it to MSIL.
```

And merge this sentence with the next?

> This gives incredibly improved runtime performance, application startup times and developers will get syntax errors at build time.

Reviewer: "incredibly improved" is maybe a bit clumsy?

```suggestion
This greatly improves runtime performance, application startup times and allows errors to be caught at build time instead of runtime.
```

> This also allows debugging of XAML code, where you can step through XAML as your application runs.

Reviewer: Starting two paragraphs with "This".

# OSX / Linux Native Menus
In order for Avalonia applications to feel properly at home on OSX we have implemented the apis to allow native menu integration.

> OSX places a menu that is permanently at the top of the screen. The first item of which the title matched the application name, is a global application menu, and is displayed regardless of the selected window or if the application is minimised.

Reviewer: Feels like if we're going to explain what OSX menus are that should be the first thing.

Avalonia will now create this menu for you, and generate the standard menu items.
![OSXAppMenu](/blog/2019-12-13-avalonia-0.9.0-release/osx-app-menu.png)

If you want to customise this menu, then you can do so by declaring a Native Menu in your Application.xaml file like so.

```xml
<Application>
  <NativeMenu.Menu>
    <NativeMenu>
      <NativeMenuItem Header="About MyApp" Command="{Binding AboutCommand}" />
    </NativeMenu>
  </NativeMenu.Menu>
</Application>
```

The consequence of this is that the Application class now has a DataContext property, where you can assign a ViewModel and bind to properties, just like any other control in Avalonia.

The Window menus of course are window specific, and can be declared like:

```xml

> <NativeMenu.Menu>

Reviewer: Maybe wrap this in `<Window>` as the application example has that element?

    <NativeMenu>
      <NativeMenuItem Header="File">
        <NativeMenuItem.Menu>
          <NativeMenu>
            <NativeMenuItem Header="Open" Command="{Binding OpenCommand}"/>
            <NativeMenuItemSeperator/>
            <NativeMenuItem Header="Recent">
              <NativeMenuItem.Menu>
                <NativeMenu/>
              </NativeMenuItem.Menu>
            </NativeMenuItem>
            <NativeMenuItemSeperator/>
            <NativeMenuItem Header="Quit Avalonia" Command="{Binding CloseCommand}" Gesture="CMD+Q"/>
          </NativeMenu>
        </NativeMenuItem.Menu>
      </NativeMenuItem>
      <NativeMenuItem Header="Edit">
        <NativeMenuItem.Menu>
          <NativeMenu>
            <NativeMenuItem Header="Copy"/>
            <NativeMenuItem Header="Paste"/>
          </NativeMenu>
        </NativeMenuItem.Menu>
      </NativeMenuItem>
    </NativeMenu>
  </NativeMenu.Menu>
```

As you can see, native menus support gestures just like a normal MenuItem would.

![OSXWindowMenu](/blog/2019-12-13-avalonia-0.9.0-release/osx-window-menu.png)

# Optional Managed File Dialogs

> Some operating systems may not have their own File dialogs, or you may prefer to use Avalonia own file dialog for consistent UI.

Reviewer: ```suggestion
Some operating systems may not have their own File dialogs, or you may prefer to use Avalonia's own file dialog for consistent UI.
```

![ManagedDialogs](/blog/2019-12-13-avalonia-0.9.0-release/managed-file-dialog.png)

You can configure Avalonia to use the managed dialog by adding to the appbuilder like so:

```csharp
AppBuilder.Configure<App>()
.UsePlatformDetect()
.UseManagedSystemDialogs();
```

If you wish the file dialog to be hosted inside your own window, where you control the chrome, colour scheme, title, icon, etc you can also use...

```csharp
.UseManagedSystemDialogs<AppBuilder, MyCustomWindowType>();
```

# ItemsRepeater

> The ItemsRepeater control has been ported from UWP. This gives very high performance when displaying very long lists.

Reviewer: ```suggestion
The [`ItemsRepeater`](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/items-repeater) control has been ported from UWP. This control is a high-performance virtualized list that supports many different layouts (via [Attached Layouts](https://docs.microsoft.com/en-us/windows/uwp/design/layout/attached-layouts)), smooth scrolling and grouping.

In the long-term we intend to use `ItemsRepeater` to implement Avalonia's `ItemsPresenter` so that all `ItemsControl`s can take advantage of these great features.
```

# New Grid and GridSplitter

> Now that WPF has been opensourced we have directly ported Grid and Grid Splitter from those repositories. This means you can now expect 100% identical behavior of these fundamental controls.

Reviewer: ```suggestion
In December last year, Microsoft announced that WPF would be open sourced. In the intervening months, the source code was published and we have been able to take advantage of this by porting `Grid` and `GridSplitter` from WPF. This means you can now expect 100% identical behavior between Avalonia and WPF of these fundamental controls.
```

# Performance Improvements

> Its worth mentioning that there have been many improvements to Avalonia especially to improve performance, reducing the amount of memory and improving speed. Many projects are now able to run smoothly on embedded platforms like Raspberry PI for instance.

Reviewer: ```suggestion
It's worth mentioning that there have been many improvements to Avalonia especially to improve performance, reducing the amount of memory and improving speed. Many projects are now able to run smoothly on embedded platforms like Raspberry PI.
```

# Production Ready
Many people come and ask, is Avalonia production ready? The answer is YES!

> There are now several applications built on Avalonia with quite large user bases. We see all the time many commercial products that are now porting to Avalonia.

Reviewer: ```suggestion
There are now several applications built on Avalonia with large user bases. We are starting to see a number of commercial products being ported to Avalonia, and we hope that trend will continue!
```

Reviewer 2: Probably worth mentioning the [Projects that are using Avalonia](https://github.com/AvaloniaUI/Avalonia/wiki/Projects-that-are-using-Avalonia) page, and probably providing shots of most epic apps.

The teams behinds these projects have been involved with testing this release.

For this reason we will be maintaining a branch of the 0.9 release with only bug fixes. So 0.9 will be maintained and supported even as we start the development of the next major version of Avalonia.
 

>  # Getting started

Reviewer: Mention the company here?

In addition, we're pleased to announce that we have formed a company to provide support contracts and custom development around Avalonia. The idea is to make Avalonia development sustainable in the long term: we're really pleased with what we've managed to get done with only volunteers but we feel we're at the point where we need some full-time developers on the project. We're still working out the details of pricing etc, but if you're interested drop us a line.

Follow instructions [here](/docs/quickstart).

# Special Thanks
This release has only been possible due to the dedicated work of many people, notably

> [grokys](https://github.com/grokys), [jkoritzinsky](https://github.com/jkoritzinsky), [kekekeks](https://github.com/kekekeks), [jmacato](https://github.com/jmacato), [marchingcube](https://github.com/marchingcube), [wieslawsoltes](https://github.com/wieslawsoltes), [gillibald](https://github.com/gillibald) and many others.

Reviewer: You can add yourself here, don't be modest ;)

Author: Felt a bit wierd thanking myself hehe but ok then :)

The best way to support Avalonia is to get involved, implement a feature, fix a bug or help test. See [contributing](http://avaloniaui.net/contributing/contributing) for information on how to get started.

Otherwise you can sponsor Avalonia financially via [OpenCollective](https://opencollective.com/Avalonia#sponsor).

> We hope you enjoy developing with Avalonia please let us know what you are building.

Reviewer: ```suggestion
We hope you enjoy developing with Avalonia - please let us know what you are building!
```

Title: Avalonia 0.9.0 Release
Published: 2019-12-13
Category: Release
Author: Dan Walmsley
---

We are pleased to announce that [Avalonia](https://github.com/AvaloniaUI/Avalonia) 0.9.0 has been
released.

0.9 is a huge update, it has been extensively tested and brings some great new features and improvements.

# Compiled XAML

> In previous releases XAML was parsed at runtime. had a non-significant runtime overhead and that some errors could cause runtime crashes.

Reviewer 3: ```suggestion
In previous releases, XAML was parsed at runtime and it had significant runtime overhead as well as some errors could cause runtime crashes.
```

> For 0.9 we have built a XAML compiler that parses XAML at build time, and compiles it to MSIL.

Reviewer 3: ```suggestion
For this release, we have built a XAML compiler that parses XAML at build time and compiles it to MSIL.
```

This greatly improves runtime performance, application startup times and allows errors to be caught at build time instead of runtime.

The new XAML compiler also allows the debugger to step into XAML code.

# OSX / Linux Native Menus
OSX places a menu that is permanently at the top of the screen. The first item of which the title matches the application name, it is a global application menu, and is displayed regardless of the selected window or if the application is minimised.

> For applications feel like they belong on OSX they need to implement these menus. The same applies to some window managers on Linux, although it is less common.

Reviewer 3: ```suggestion
For applications to feel like they belong on OSX, they would need to implement these menus. The same applies to some window managers on Linux, although it is less common.
```

> Avalonia will now create this menu for you, and generate the standard menu items that OSX users expect.

Reviewer 3: menu "layout" flows better in this i think
```suggestion
Avalonia will now create this menu for you, and generate the standard menu layout that OSX users expect.
```

Author: I think layout implies structure, where as here we are referring to 4 or 5 items on the App menu, like hide other windows, quit, etc

Some operating systems may not have their own File dialogs, or you may prefer to use Avalonia's own file dialog for consistent UI.

![ManagedDialogs](/blog/2019-12-13-avalonia-0.9.0-release/managed-file-dialog.png)

You can configure Avalonia to use the managed dialog by adding to the appbuilder like so:

```csharp
AppBuilder.Configure<App>()
.UsePlatformDetect()
.UseManagedSystemDialogs();
```

If you wish the file dialog to be hosted inside your own window, where you control the chrome, colour scheme, title, icon, etc you can also use...

```csharp
.UseManagedSystemDialogs<AppBuilder, MyCustomWindowType>();
```

# ItemsRepeater
The [`ItemsRepeater`](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/items-repeater) control has been ported from UWP. This control is a high-performance virtualized list that supports many different layouts (via [Attached Layouts](https://docs.microsoft.com/en-us/windows/uwp/design/layout/attached-layouts)), smooth scrolling and grouping.

In the long-term we intend to use `ItemsRepeater` to implement Avalonia's `ItemsPresenter` so that all `ItemsControl`s can take advantage of these great features.

# New Grid and GridSplitter

> Now that WPF has been opensourced we have directly ported Grid and Grid Splitter from those repositories. This means you can now expect 100% identical behavior of these fundamental controls.

Reviewer 3: Duplicate line?
